### ISSUE TITLE
fix(pytorch): in-place mutation of B_diag and S_corr breaks torch.compile and risks gradient corruption

### ISSUE THREAD

**Issue by hunter-heidenreich** [hunter-heidenreich/Kabsch-Cookbook#151]
## Summary

`kabsch()` and `kabsch_umeyama()` in the PyTorch backend use in-place indexed assignment to build the determinant-sign correction diagonal. This pattern can corrupt autograd gradients and breaks `torch.compile` tracing.

## Locations

- `src/kabsch_horn/pytorch/kabsch_svd_nd.py:187` — `B_diag[..., -1] = d_sign`
- `src/kabsch_horn/pytorch/kabsch_svd_nd.py:295` — `S_corr[..., -1] = d_sign`

## Problem

`B_diag` is created with `torch.ones(...)` and then mutated in-place via `B_diag[..., -1] = d_sign`. Because `d_sign` depends on `V` and `U` through `torch.linalg.det`, this tensor participates in the autograd graph. The in-place write:

1. **Breaks `torch.compile`** — the Dynamo tracer cannot trace through in-place indexed assignment on tensors in the graph.
2. **Can corrupt gradients** — PyTorch's autograd engine may need the pre-mutation version of the tensor for backward, and the in-place write invalidates it.

The same pattern appears in `kabsch_umeyama()` with `S_corr`.

## Suggested fix

Replace with out-of-place construction, e.g.:

```python
B_diag = torch.cat(
    [torch.ones(*d_sign.shape, D - 1, dtype=P.dtype, device=P.device),
     d_sign.unsqueeze(-1)],
    dim=-1,
)
```

Or use `torch.where` with a last-index boolean mask. Apply the same fix to `S_corr` in `kabsch_umeyama()`.

## Found by

PyTorch framework audit (autograd correctness + torch.compile compatibility)
